### DISCUSSION TITLE
Enumerate all possible R-group decompositions

### DISCUSSION THREAD

**Discussion by jpfeuffer** [rdkit/rdkit#7297] [Q&A]
Hi!

I would like to enumerate all possible R-group decompositions for a given core, not only one. Is this possible with the built-in functionality?

![image](https://github.com/rdkit/rdkit/assets/8102638/2c838d52-a17d-4c1b-8bd1-13747e51646f)

Core (rgd_test.sdf):

```

  Mrv2311 03262410502D          

 12 10  0  0  1  0            999 V2000
   -3.2067    0.3583    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.4943   -0.0577    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -1.7778    0.3512    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.7736    1.1762    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    2.9221   -0.0243    0.0000 R#  0  0  0  0  0  0  0  0  0  0  0  0
    1.4976   -0.8573    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    2.2054    0.3841    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.4933   -0.0323    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.7764    0.3762    0.0000 N   0  0  0  0  0  3  0  0  0  0  0  0
    0.4018    1.1113    0.0000 R#  0  0  0  0  0  0  0  0  0  0  0  0
    0.1930   -0.2072    0.0000 R#  0  0  0  0  0  0  0  0  0  0  0  0
   -1.0654   -0.0649    0.0000 R#  0  0  0  0  0  0  0  0  0  0  0  0
  2  1  1  0  0  0  0
  3  2  1  0  0  0  0
  4  3  2  0  0  0  0
  7  5  1  0  0  0  0
  8  7  1  0  0  0  0
  9  8  1  0  0  0  0
  9 10  1  0  0  0  0
  9 11  1  0  0  0  0
  8  6  2  0  0  0  0
  3 12  1  0  0  0  0
M  RGP  4   5   5  10   2  11   3  12   1
M  END
$$$$
```

```python

ms = [x for x in Chem.SDMolSupplier('./rgd_test.sdf')]
ms += [Chem.MolFromSmiles('[H]OC(=O)OC1=CC(=CN(C1)C(=O)OCCC)C(=O)O[H]')]
res,_ = rdRGroupDecomposition.RGroupDecompose([ms[0]],ms[1:])
for row in res:
    print(f'  {row}')

Draw.MolsToGridImage(res[0].values(),legends=res[0].keys())
```

Result: Only one decomposition
Expected result: My assumption is, that two are possible (for different matches of R1).

I tried some RGD params but none worked so far.

I know it's probably not your typical decomp but it would make my life much easier if this were possible.

Thanks

**Comment by bp-kelley** [comment]
The RGroup decomposition will only give you one match here.  It is designed
to keep the same rgroup fragments together as much as possible when doing a
decomposition, in this case since you had only one molecule, it will be an
arbitrary match.

Also note that the decomposition will look weird because you are splitting
a ring.  I have some code that makes it prettier I can post in a bit.

**Reply by jpfeuffer** [reply]
I see. Yes I noticed the "linearization" of the rings already.
Unfortunately, I would really need all matches to be able to use that functionality.
I want to enumerate and transfer the matched groups to another molecule and create multiple versions of it.

**Comment by bp-kelley** [comment]
What do you mean by "all possible r-group decompositions"?

Internally the code has the option to enumerate all symmetry matches and then scores the decompositions to pick one that best fits the entire set.  Would you like this exposed or simply know how to do it?

**Reply by jpfeuffer** [reply]
Thanks a lot for your help already.

From how I understand symmetry matches, this should not be a problem for me.
My main problem is the fact that different components of the R group query may match to different places in the "molecule to decompose" and therefore I want to enumerate all combinations of those matches and the resulting R-group decompositions of those.

In my initial example I get:
![image](https://github.com/rdkit/rdkit/assets/8102638/a6dcd78f-9313-43cd-bfe8-2ffe6a074b6f)

I would like a second result for the mapping of the carboxyl to the more distant group (ix: 0,1,2 in the OP).

Ideally it would be cool to be exposed from C++ to python but if you have a good pseudocode, I can try reimplementing it in python if the exposure might take some time.

**Comment by jakecleghorn** [comment]
Hi,

I came across this discussion earlier this year when looking for a solution to the same issue. In RDKit versions ≥ 2025.09.2, it's now possible to handle these cases with the `allowMultipleCoresInSameMol` parameter in `RGroupDecompositionParameters`.

Using your example:

```python
params = rdRGroupDecomposition.RGroupDecompositionParameters()
params.allowMultipleCoresInSameMol = True

rgd_test = Chem.MolFromMolBlock(rgd_test_molblock)   # where rgd_test_molblock is the molblock from your rgd_test.sdf

ms = [Chem.MolFromSmiles('[H]OC(=O)OC1=CC(=CN(C1)C(=O)OCCC)C(=O)O[H]')]
res,_ = rdRGroupDecomposition.RGroupDecompose(rgd_test, ms, options=params)
for row in res:
    print(f'  {row}')
```

This gives the two possible decompositions:

### res[0]
<img width="474" height="315" alt="image" src="https://github.com/user-attachments/assets/fdc704f2-f96b-40f2-8954-fe6b2907febf" />


### res[1]
<img width="490" height="332" alt="image" src="https://github.com/user-attachments/assets/42c8fd1c-197b-4cdc-88e8-5c519861b110" />

**Reply by jpfeuffer** [reply]
Very nice! Thank you for replying here! That is exactly what I was looking for.
